test/js_test.go

```diff
@@ -955,6 +955,381 @@ func TestJetStreamSubscribe(t *testing.T) {
 	}
 }
 
+func TestPullSubscribeFetchBatch(t *testing.T) {
+	s := RunBasicJetStreamServer()
+	defer shutdownJSServerAndRemoveStorage(t, s)
+
+	nc, js := jsClient(t, s)
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"foo"},
+	})
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	t.Run("basic fetch", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 5; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		res, err := sub.FetchBatch(10)
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			for i := 0; i < 5; i++ {
+				js.Publish("foo", []byte("msg"))
+			}
+		}()
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() != nil {
+			t.Fatalf("Unexpected error: %s", res.Error())
+		}
+		if len(msgs) != 10 {
+			t.Fatalf("Expected %d messages; got: %d", 10, len(msgs))
+		}
+	})
+
+	t.Run("multiple concurrent fetches", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 50; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		var r1, r2, r3 nats.MessageBatch
+		started := &sync.WaitGroup{}
+		started.Add(3)
+		errs := make(chan error, 3)
+		go func() {
+			r1, err = sub.FetchBatch(10)
```

Comment:
this `err` being shared among goroutines might seem to be causing a data race in the test

```
=== RUN   TestPullSubscribeFetchBatch/multiple_concurrent_fetches
==================
WARNING: DATA RACE
Write at 0x00c0003b64f8 by goroutine 50:
  github.com/nats-io/nats%2ego.(*Subscription).FetchBatch.func2()
     nats-dev/src/github.com/nats-io/nats.go/js.go:2974 +0x444

Previous read at 0x00c0003b64f8 by goroutine 42:
  github.com/nats-io/nats%2ego.(*messageBatch).Done()
      nats-dev/src/github.com/nats-io/nats.go/js.go:2791 +0x30
  github.com/nats-io/nats.go/test.TestPullSubscribeFetchBatch.func2()
      nats-dev/src/github.com/nats-io/nats.go/test/js_test.go:1074 +0xc14
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.19.5/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.19.5/libexec/src/testing/testing.go:1493 +0x40

Goroutine 50 (running) created at:
  github.com/nats-io/nats%2ego.(*Subscription).FetchBatch()
      nats-dev/src/github.com/nats-io/nats.go/js.go:2945 +0xf04
  github.com/nats-io/nats.go/test.TestPullSubscribeFetchBatch.func2.1()
      nats-dev/src/github.com/nats-io/nats.go/test/js_test.go:1022 +0x64

Goroutine 42 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/Cellar/go/1.19.5/libexec/src/testing/testing.go:1493 +0x55c
  github.com/nats-io/nats.go/test.TestPullSubscribeFetchBatch()
      nats-dev/src/github.com/nats-io/nats.go/test/js_test.go:1006 +0x368
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.19.5/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.19.5/libexec/src/testing/testing.go:1493 +0x40
==================
    js_test.go:1075: FetchBatch result channel should be closed after receiving all messages on r1
    testing.go:1319: race detected during execution of test
```